### PR TITLE
Issue 471/edit view title

### LINKED ISSUE #471: Edit view title
## Description
A user should be able to change the name of a view by editing it in place on the page.

## Relevant job stories
* When I'm creating a new view, I want to give it a name so that I can easily find it later

## Prerequisites
This task can be started once #494 has been completed.

## Requirements 
- [ ] Hovering the title should show a dotted outline and a tooltip saying "Change the title"
- [ ] Clicking the title should enter edit mode, where user can type to edit the title
  - [ ] If the title is the default one (localized version of "Untitled view") reset to empty when clicked
- [ ] Clicking outside or pressing enter should save the new title
  - [ ] If empty, revert back to default title (localized version of "Untitled view")

## Possible implementations
Either switch to a MUI input when clicked, or use `contentEditable`?

Relevant API resources:
* `/orgs/{org_id}/people/views/{view_id}` (`PATCH` to change title)

## Design specifications
See wireframe for context, although the exact flow described above has not been wireframed in detail.

image

## Open questions
No outstanding questions at this time.

### PR DESCRIPTION
**Pull request by lowlandjuju**
Fixes #471 

Implements everything in #471, and includes unit and integration tests, plus some improvised animation and edge case handling.

issue-471

### CODE REVIEW

File: src/fetching/views/patchView.ts
Comment (djbusstop) [inline src/fetching/views/patchView.ts:15]: We have an `APIError` for this.

File: src/components/layout/organize/SingleViewLayout.tsx
Comment (djbusstop) [inline src/components/layout/organize/SingleViewLayout.tsx:26]: Here you could use NProgress to show the loading state of the update, and you could use an MUI Snackbar to show an error if it fails.

Comment (lowlandjuju) [reply src/components/layout/organize/SingleViewLayout.tsx:26]: - The update is IMO not slow enough to warrant a progress bar, and there are enough visual cues already if you try the component out
- There's a snackbar is in the EditTextInPlace component - we could move it to the parent component, but I don't see a good reason. Think this kind of thing should probably be handled centrally by some notification provider, and some consistent way of display notifications, but this is a quick short term solution.

File: src/fetching/views/patchView.ts
Comment (lowlandjuju) [reply src/fetching/views/patchView.ts:15]: I'll take a look! I just copied and pasted this code block from patchCampaign for my sins....

**Comment by richardolsson** [comment]
Awesome! Before review, could you maybe provide a brief description (and perhaps even a screenshot) as well as a link to the issue it resolves?

**Comment by lowlandjuju** [comment]
@richardolsson went one better than a screenshot 👍

File: playwright/tests/organize/people/views/view-detail.spec.ts
Comment (djbusstop) [inline playwright/tests/organize/people/views/view-detail.spec.ts:85]: Beautiful!

File: src/fetching/views/patchView.ts
Comment (djbusstop) [inline src/fetching/views/patchView.ts:6]: The type for the `view` argument would be a `Partial<ZetkinView>`.

File: src/fetching/views/patchView.ts
Comment (djbusstop) [inline src/fetching/views/patchView.ts:18]: All Zetkin API requests that return something put it in to the property `data`. Most of our fetch handlers are like this
```js
const resBody = await res.json();
return resBody.data
```

Because of this the return type isn't correct, the return type presently would be `Promise<{ data: ZetkinView}>`

Even though you aren't using this handler in any case where you need to return this value, for consistency's sake, and for types' sakes, I would say it's best to return the `resBody.data`.

File: src/components/layout/organize/SingleViewLayout.tsx
Comment (djbusstop) [inline src/components/layout/organize/SingleViewLayout.tsx:27]: In `src/people/views/[viewId].tsx` that the query key used in `prefetchQuery` is `['view', viewId]`. Since that's the query we want to invalidate, we should use the same key. Just need to add `viewId` to the list!

Comment (djbusstop) [reply src/components/layout/organize/SingleViewLayout.tsx:27]: Also, I don't think we want to use try / catch here, and I don't think the return true / false is needed because it doesn't look like you're using the boolean state for anything. 

In the case you wanted to use these booleans to show an error message to the user if the handler threw an error, you would instead be better off using the `onSuccess` and `onError` properties in the same way you used `onSettled.

Comment (djbusstop) [reply src/components/layout/organize/SingleViewLayout.tsx:27]: Finally, this method doesn't need to be async. Since it's just going to resolve and invalidate a query. I would simply write it like this:

```js
const updateTitle = (view: ZetkinView, newTitle: string) => {
  patchViewMutation.mutate({ title: newTitle }, {
    onSettled: () => queryClient.invalidateQueries(['view' ]),
  })
};
```

File: src/components/EditTextInPlace/index.tsx
Comment (djbusstop) [inline src/components/EditTextInPlace/index.tsx:81]: Sorry I realise that here is where the boolean state of the onSubmit it being used. I don't think it's a great pattern to use booleans for handling these error states so I would think about a different way to handle this... I don't know if the child handling the error of a function passed in from the parent is a good method in any case. Maybe @richardolsson has another opinion and thinks this is a good idea so interested what he thinks.

Comment (richardolsson) [reply src/components/EditTextInPlace/index.tsx:81]: Oh I just commented on that other thread about this. But yeah, I had that same concern when me and Everin first talked about it, and again when we talked about it all three of us, but then I dropped it because I realized this would be refactored with the stuff you're working on in #514 anyway.

I would be very happy to see this change, so that the error is handled outside (`onError` in the mutation, as you suggested).

File: src/components/layout/organize/SingleViewLayout.tsx
Comment (djbusstop) [reply src/components/layout/organize/SingleViewLayout.tsx:27]: Yes, I realised that. But I think if this is a generic component that we would use everywhere, I don't think it's ideal to have to pass a function that returns a boolean in such a way, it enforces too much structure on the `onSubmit` handler, and doesn't use try/catch internally. I wonder if I'm using the the word "antipattern" correctly if I say that's what it looks like?

File: src/components/EditTextInPlace/index.tsx
Comment (djbusstop) [reply src/components/EditTextInPlace/index.tsx:81]: Even in my refector I think this wouldn't be fixed, because wherever the mutation would be implemented would require a return of `boolean` so I think it's worth figuring out a better way hmmmmm.

I wonder if the best way is to have an `error` prop which can be set from the parent. Then one could do `error={patchViewMutation.isError}`

Comment (richardolsson) [reply src/components/EditTextInPlace/index.tsx:81]: My thought about the refactor is that errors will probably be handled inside the code generated by the `createMutation()` factory, so we could then take out the error handling code entirely from this specific implementation.

But yeah, I agree that it should be refactored here, now. And I think the way to do it is to trigger the snackbar from `SingleViewLayout` rather than inside `EditTextInPlace`. Once we decide how to communicate errors to the user, I think that will be easier to implement.

If by then we want snackbars to be triggered consistently in response to errors, that can be handled in the factory. If we want snackbars only in this case, then doing it in `SingleViewLayout` is the right way anyway. And if we don't want snackbars at all, we can remove them here without risking breaking the `EditTextInPlace` which may then be used elsewhere.

**Comment by djbusstop** [comment]
The only change I made that is noticable from a UI perspective is that I moved the error message that says "This cannot be empty" and put it in the toolbar when the input is empty. This is because the padding was a bit odd before for the error message, and because it still said "press enter to save" when empty. Now all messages users need to see are in the toolbar :+1: Hope this is OK!

File: src/components/layout/organize/SingleViewLayout.tsx
Comment (richardolsson) [inline src/components/layout/organize/SingleViewLayout.tsx:55]: …that shows, what? 😊

**Review by richardolsson — CHANGES_REQUESTED** [review]
I think this is looking awesome! Great work, both of you. 🙏  I found some bugs though, that I think we should try to fix before merging.

## Text cut off
For some values, the element is are not made wide enough, so the text is cut off. Example:
image

## Not reset between views
Try navigating directly between two views (using the jump menu). The title should update to reflect the separate views, but doesn't because the old title is in state. Steps to reproduce:

1. Navigate to a view, A
2. Edit title
3. Use jump menu to navigate to another view, B

Expected results: Component should display title of view B
Actual results: Components displays title of view A

The simplest solution to this IMO is to add a `key` property to where the `EditTextInPlace` component is used in `SingleViewLayout`. Use something that is unique for each view, e.g. it's `view.id`. That would solve the problem, since it would tell React that the component instance is actually not the same, even though it's the same component in the same place in the tree. That will force it to re-mount, resetting state.

**Comment by richardolsson** [comment]
Seems like the same problem exists in the opposite direction as well. 😞 

image

File: src/components/layout/organize/SingleViewLayout.tsx
Comment (djbusstop) [reply src/components/layout/organize/SingleViewLayout.tsx:55]: lolllll

**Comment by djbusstop** [comment]
> Seems like the same problem exists in the opposite direction as well. disappointed
> 
> image

Sorry I don't understand what you mean by the problem exists in the opposite direction?

**Comment by djbusstop** [comment]
The size issue is one I'm very confused by, I am trying to fix it but it's tedious. The `size` prop on the input I think is meant for monotype fonts...

**Comment by richardolsson** [comment]
> Sorry I don't understand what you mean by the problem exists in the opposite direction?

I mean the field is too wide, i.e. it grows faster than the text contained grows as you're typing.

**Review by richardolsson — APPROVED** [review]
Great work team! I've made the last couple of improvements myself, and looked through the rest of the code as well. I think this turned out really nice! 🙌 

I think I'll merge this now since the PRs are starting to pile up. 🎉

**Pull request merged**
